kjdoc/tour.md

```diff
@@ -538,13 +538,440 @@ This section describes KJ APIs that control process execution and low-level inte
+});
+```
+
+If a promise rejects (throws an exception), then the exception propagates through `.then()` to the new derived promise, without calling the continuation. If you'd like to actually handle the exception, you may pass a second lambda as the second argument to `.then()`.
+
+```c++
+kj::Promise<kj::String> promise = networkAddress.connect()
+    .then([](kj::Own<kj::AsyncIoStream> stream) {
+  return stream->readAllText().attach(kj::mv(stream));
+}, [](kj::Exception&& exception) {
+  return kj::str("connection error: ", exception);
+});
+```
+
+You can also use `.catch_(errorHandler)`, which is a shortcut for `.then(identityFunction, errorHandler)`.
+
+### `kj::evalNow()`, `kj::evalLater()`, and `kj::evalLast()`
+
+These three functions take a lambda as the parameter, and return the result of evaluating the lambda. They differ in when, exactly, the execution happens.
+
+```c++
+kj::Promise<int> promise = kj::evalLater([]() {
+  int i = doSomething();
+  return i;
+});
+```
+
+As with `.then()` continuations, the lambda passed to these functions may itself return a `Promise`.
+
+`kj::evalNow()` executes the lambda immediately -- before `evalNow()` even returns. The purpose of `evalNow()` is to catch any exceptions thrown and turn them into a rejected promise. This is often a good idea when you don't want the caller to have to handle both synchronous and asynchronous exceptions -- wrapping your whole function in `kj::evalNow()` ensures that all exceptions are delivered asynchronously.
+
+`kj::evalLater()` executes the lambda on a future turn of the event loop. This is equivalent to `kj::Promise<void>().then()`.
+
+`kj::evalLast()` arranges for the lambda to be called only once all other work queued to the event loop has completed (but before querying the OS for new I/O events). This can often be useful, for example, for batching. For example, if a program tends to make many small write()s to a socket in rapid succession, you might want to add a layer that collects the writes into a batch, then sends the whole batch in a single write from an `evalLast()`. This way, none of the bytes are significantly delayed, but they can still be coalesced.
+
+If multiple `evalLast()`s exist at the same time, they will execute in last-in-first-out order. If the first one out schedules more work on the event loop, that work will be completed before the next `evalLast()` executes, and so on.
+
+### Attachments
+
+Often, a task represented by a `Promise` will require that some object remains live until the `Promise` completes. In particular, under KJ conventions, unless documented otherwise, any class method which returns a `Promise` inherently expects that the caller will ensure that the object it was called on will remain live until the `Promise` completes (or is canceled). Put another way, member function implementations may assume their `this` pointer is valid as long as their returned `Promise` is alive.
```

Comment:
```suggestion
Often, a task represented by a `Promise` will require that some object remains alive until the `Promise` completes. In particular, under KJ conventions, unless documented otherwise, any class method which returns a `Promise` inherently expects that the caller will ensure that the object it was called on will remain alive until the `Promise` completes (or is canceled). Put another way, member function implementations may assume their `this` pointer is valid as long as their returned `Promise` is alive.
```

Reply:
Hmm, I sort of like "live" better here, but don't have any particularly good argument.

Reply:
Not a big deal... I was thinking to myself that "alive" was a little better because:
- "live" is more ambiguous, in that it can mean either "alive" or "in active use"
- the paragraph already also uses "alive" to refer to the promise, which might lead someone to question the distinction.